### DISCUSSION TITLE
https://ghcr.io/v2 :UNAUTHORIZED, authentication required

### DISCUSSION THREAD

**Discussion by karawitan** [community/community#35172] [Actions]
Hello,

  while accessing the following url :

https://ghcr.io/v2/homebrew/core/go/manifests/1.19.1

I get

{"errors":[{"code":"UNAUTHORIZED","message":"authentication required"}]}

This is strange because it's a public asset included in the homebrew project.

Is there an issue with github packages ?

Best Regards,

Olivier

**Comment by airtower-luna** [comment]
That's annoying but normal. :sweat_smile: 

You still need to provide a token even for public assets, but you can get an anonymous one from the endpoint described in the `www-authenticate` header. If I make a GET request to the URL you mentioned, I get:

```
www-authenticate: Bearer realm="https://ghcr.io/token",service="ghcr.io",scope="repository:homebrew/core/go:pull"
```

Make a request to the URL in `realm` with `service` and `scope` as URL parameters, in this case:

```
https://ghcr.io/token?service=ghcr.io&scope=repository%3Ahomebrew/core/go%3Apull
```

You should get a JSON object containing a `token` field. Use the value of that like you would use a PAT (or other OAuth token). E.g. (token shortened, replace with the one you got):

```
curl -v -H 'Authorization: Bearer djE6...' -H 'Accept: application/vnd.oci.image.index.v1+json' https://ghcr.io/v2/homebrew/core/go/manifests/1.19.1
```

**Comment by maltfield** [comment]
In Debian linux, here's some easier ways to get the token and download the homebrew manifest for the `wget` package

```
# install depends
sudo apt-get install curl jq

# get a JSON with an anonymous token
curl -so "token.json" "https://ghcr.io/token?service=ghcr.io&scope=repository%3Ahomebrew/core/go%3Apull";

# extract token from JSON
token=$(cat token.json | jq -jr ".token")

# download the manifest file
curl -o manifest.json -v -H "Authorization: Bearer ${token}" -H 'Accept: application/vnd.oci.image.index.v1+json' https://ghcr.io/v2/homebrew/core/wget/manifests/1.21.3

# view the downloaded file
cat manifest.json
```

See also:
 * https://apple.stackexchange.com/questions/470937/how-to-get-the-url-to-download-a-homebrew-bottle

Update: Apparently you actually don't need to fetch a token. You can just use `QQ==`

```
# download the manifest file
curl -o manifest.json -v -H "Authorization: Bearer QQ==" -H 'Accept: application/vnd.oci.image.index.v1+json' https://ghcr.io/v2/homebrew/core/wget/manifests/1.21.3

# view the downloaded file
cat manifest.json
```

For more information, see https://tech.michaelaltfield.net/2024/09/03/container-download-curl-wget/
